daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Upgrading.scala

```diff
@@ -0,0 +1,464 @@
+  private lazy val packageId: Upgrading[Ref.PackageId] = packagesAndIds.map(_._1)
+  private lazy val _package: Upgrading[Ast.Package] = packagesAndIds.map(_._2)
+
+  private def extractDelExistNew[K, V](
+      arg: Upgrading[Map[K, V]]
+  ): (Map[K, V], Map[K, Upgrading[V]], Map[K, V]) =
+    (
+      arg.past -- arg.present.keySet,
+      arg.past.keySet
+        .intersect(arg.present.keySet)
+        .map(k => k -> Upgrading(arg.past(k), arg.present(k)))
+        .toMap,
+      arg.present -- arg.past.keySet,
+    )
+
+  private def checkDeleted[K, V, T <: Throwable](
+      arg: Upgrading[Map[K, V]],
+      handler: (K, V) => T,
+  ): Try[(Map[K, Upgrading[V]], Map[K, V])] = {
+    val (deletedV, existingV, newV) = extractDelExistNew(arg)
+    deletedV.headOption match {
+      case Some((k, v)) => Failure(handler(k, v))
+      case _ => Success((existingV, newV))
+    }
+  }
+
+  private def tryAll[A, B](t: Iterable[A], f: A => Try[B]): Try[Seq[B]] =
+    Try(t.map(f(_).get).toSeq)
+
+  private def check(): Try[Unit] = {
+    for {
+      (upgradedModules, newModules @ _) <-
+        checkDeleted(
+          _package.map(_.modules),
+          (name: Ref.ModuleName, _: Ast.Module) => MissingModule(name),
+        )
+      _ <- tryAll(upgradedModules.values, checkModule(_))
+    } yield ()
+  }
+
+  private def checkModule(module: Upgrading[Ast.Module]): Try[Unit] = {
+    def datatypes(module: Ast.Module): Map[Ref.DottedName, Ast.DDataType] =
+      module.definitions.flatMap(_ match {
+        case (k, v: Ast.DDataType) => Some((k, v));
+        case _ => None;
+      })
+
+    val moduleWithMetadata = module.map(ModuleWithMetadata)
+    for {
+      (existingTemplates, _new) <- checkDeleted(
+        module.map(_.templates),
+        (name: Ref.DottedName, _: Ast.Template) => MissingTemplate(name),
+      )
+      _ <- tryAll(existingTemplates, checkTemplate(_))
+
+      (existingDatatypes, _new) <- checkDeleted(
+        module.map(datatypes(_)),
+        (name: Ref.DottedName, _: Ast.DDataType) => MissingDataCon(name),
+      )
+      _ <- tryAll(existingDatatypes, checkDatatype(moduleWithMetadata, _))
+    } yield ()
+  }
+
+  private def checkTemplate(
+      templateAndName: (Ref.DottedName, Upgrading[Ast.Template])
+  ): Try[Unit] = {
+    val (templateName, template) = templateAndName
+    for {
+      (existingChoices, _newChoices) <- checkDeleted(
+        template.map(_.choices),
+        (name: Ref.ChoiceName, _: Ast.TemplateChoice) => MissingChoice(name),
+      )
+
+      _ <- tryAll(existingChoices.values, checkChoice(_))
+      _ <- checkKey(templateName, template.map(_.key))
+    } yield ()
+  }
+
+  private def checkType(typ: Upgrading[Ast.Type]): Boolean = {
+    AlphaEquiv.alphaEquiv(unifyPackageIds(typ.past), unifyPackageIds(typ.present))
+  }
+
+  // TODO: Consider whether we should strip package ids from all packages in the
+  // upgrade set, not just within the pair
```

Comment:
Does this need a ticket? Can you give an example of when this is important?

Reply:
fixed